### PR TITLE
Return density_ve results by value

### PR DESCRIPTION
**Pull request by sekelle**

### CODE REVIEW

File: sph/include/sph/kernel_ve/density_kern.hpp [hunk 1]
```diff
@@ -56,10 +55,10 @@ CUDA_DEVICE_HOST_FUN inline void densityJLoop(int i, T sincIndex, T K, const cst
     T wrho0i = wrho0[i];
     T xmassi = m[i] / rho0i;
 
-    T hInv  = 1.0 / hi;
+    T hInv  = T(1) / hi;
     T h3Inv = hInv * hInv * hInv;
 
-    T kxi      = 0.0;
+    T kxi      = xmassi;
```
Comment (rmcabezon) [inline sph/include/sph/kernel_ve/density_kern.hpp:61]: Add a comment here before these kxi initialization: //self-contribution initialization. We can have a similar structure in other Jloops.

File: sph/include/sph/kernel_ve/density_kern.hpp [hunk 2]
```diff
@@ -56,10 +55,10 @@ CUDA_DEVICE_HOST_FUN inline void densityJLoop(int i, T sincIndex, T K, const cst
...
     T whomegai = 0.0;
```
Comment (rmcabezon) [inline sph/include/sph/kernel_ve/density_kern.hpp:62]: You can do the same with whomegai:
whomegai = - T(3) * xmassi

File: sph/include/sph/kernel_ve/density_kern.hpp [hunk 3]
```diff
@@ -69,20 +68,24 @@ CUDA_DEVICE_HOST_FUN inline void densityJLoop(int i, T sincIndex, T K, const cst
...
 
-    kx[i]    = K * (kxi + xmassi) * h3Inv;
-    whomegai = K * (whomegai - 3.0 * xmassi) * h3Inv * hInv;
+    kxi *= K * h3Inv;
 
-    ro[i]    = kx[i] * m[i] / xmassi;
-    whomegai = whomegai * m[i] / xmassi + (ro[i] / rho0i - K * xmassi * h3Inv) * wrho0i;
-    T dhdrho = -hi / ro[i] / 3.0; // This /3 is the dimension hard-coded.
-    gradh[i] = 1.0 - dhdrho * whomegai;
+    whomegai = K * (whomegai - T(3) * xmassi) * h3Inv * hInv;
```
Comment (rmcabezon) [inline sph/include/sph/kernel_ve/density_kern.hpp:80]: If the self-contribution initialization for whomegai is done before the loop, you can change this to:
whomegai *= K * h3Inv * hInv

File: sph/include/sph/kernel_ve/density_kern.hpp [hunk 4]
```diff
@@ -69,20 +68,24 @@ CUDA_DEVICE_HOST_FUN inline void densityJLoop(int i, T sincIndex, T K, const cst
...
+    whomegai = whomegai * rho0i + (kxi - K * xmassi * h3Inv) * wrho0i;
+
+    T dhdrho = -hi / (kxi * rho0i * T(3)); // This /3 is the dimension hard-coded.
+
+    T rhoi   = kxi * rho0i;
```
Comment (rmcabezon) [inline sph/include/sph/kernel_ve/density_kern.hpp:85]: VEs are hardcoded here. It is better to keep the flexibility to be able to change the structure of the VEs in xmass.

**Review by rmcabezon — COMMENTED** [review]
The exploit of rho = kx * rho0 hardcodes the shape of the VEs. If we can save memory somewhere else would be better to keep the flexibility to simply change Xmass. Otherwise, it will be very cumbersome and prone to error to change all the equations later.

File: sph/include/sph/kernel_ve/density_kern.hpp (hunk 1 above)
Comment (sekelle) [reply sph/include/sph/kernel_ve/density_kern.hpp:61]: Will do and move to new PR.

File: sph/include/sph/kernel_ve/density_kern.hpp (hunk 2 above)
Comment (sekelle) [reply sph/include/sph/kernel_ve/density_kern.hpp:62]: same as above

File: sph/include/sph/kernel_ve/density_kern.hpp (hunk 3 above)
Comment (sekelle) [reply sph/include/sph/kernel_ve/density_kern.hpp:80]: nice simplification

File: sph/include/sph/kernel_ve/density_kern.hpp (hunk 4 above)
Comment (sekelle) [reply sph/include/sph/kernel_ve/density_kern.hpp:85]: will be addressed in new PR

**Pull request closed without merging**